siphon/simplewebservice/igra2.py

```diff
@@ -0,0 +1,431 @@
+try:
+    from urllib.request import urlopen
+except ImportError:
+    from urllib2 import urlopen
+    
+warnings.filterwarnings('ignore', 'Pandas doesn\'t allow columns to be created', UserWarning)
+
+
+class IGRAUpperAir:
+    """Download and parse data from NCEI's Integrated Radiosonde Archive version 2."""
+    
+    def __init__(self):
+        """Set ftp site address and file suffix based on desired dataset."""
+        self.ftpsite = 'ftp://ftp.ncdc.noaa.gov/pub/data/igra/'
+        self.suffix = ''
+        self.begin_date = ''
+        self.end_date = ''
+        self.site_id = ''
+
+    @classmethod
+    def request_data(cls, time, site_id, derived=False):
+        """Retreive IGRA version 2 data for one station.
+
+        Parameters
+        --------
+        site_id : str
+            11-character IGRA2 station identifier.
+
+        time : datetime
+           The date and time of the desired observation. If list of two times is given,
+           dataframes for all dates within the two dates will be returned.
+
+        Returns
+        -------
+            :class: `pandas.DataFrame` containing the data.
+        """
+
+        igra2 = cls()
+
+        # Set parameters for data query
+        if derived:
+            igra2.ftpsite = igra2.ftpsite + 'derived/derived-por/'
+            igra2.suffix = igra2.suffix + '-drvd.txt'
+        else:
+            igra2.ftpsite = igra2.ftpsite + 'data/data-por/'
+            igra2.suffix = igra2.suffix + '-data.txt'
+
+        if type(time) == datetime.datetime:
+            igra2.begin_date = time
+            igra2.end_date = time
+        else:
+            igra2.begin_date, igra2.end_date = time
+
+        igra2.site_id = site_id
+
+        df, headers = igra2._get_data()
+
+        return df, headers
+
+    def _get_data(self):
+        """Process the IGRA2 text file for observations at site_id matching time.
+
+        Return:
+        -------
+            :class: `pandas.DataFrame` containing the body data.
+            :class: `pandas.DataFrame` containing the header data.
+        """
+        # Split the list of times into begin and end dates. If only
+        # one date is supplied, set both begin and end dates equal to that date.
+
+        body, header, dates_long, dates = self._get_data_raw()
+
+        params = self._get_fwf_params()
+
+        df_body = pd.read_fwf(StringIO(body), **params['body'])
+        df_header = pd.read_fwf(StringIO(header), **params['header'])
+        df_body['date'] = dates_long
+
+        df_body = self._clean_body_df(df_body)
+        df_header = self._clean_header_df(df_header)
+        df_header['date'] = dates
+
+        return df_body, df_header
+
+    def _get_data_raw(self):
+        """Download observations matching the time range.
+
+        Returns a tuple with a string for the body, string for the headers,
+        and a list of dates.
+        """
+        with urlopen(self.ftpsite + self.site_id + self.suffix + '.zip') as url:
```

Comment:
In Python 2, apparently you can't use the result from `urlopen` as a context manager 🙄 . You can work around this by adding (at the top):
```python
from contextlib import closing
```
and then here do:
```python
with closing(urlopen(...)) as url:
```

Reply:
Ah, glad you noticed that. I almost never use Python 2 anymore.

Reply:
Glad to hear it. I long for the day that we can drop Python 2 support.